### PR TITLE
Switches `URL` polyfill to `whatwg-url`

### PR DESCRIPTION
**Pull request by mhassan1**
This PR replaces the `URL` polyfill with `whatwg-url`. It uses `browserify` to bundle and transpile `whatwg-url`.

Resolves https://github.com/mrhenry/polyfill-library/issues/53.

### CODE REVIEW

File: polyfills/Symbol/polyfill.test.js [hunk 1]
```diff
@@ -5,7 +5,8 @@ it('is a function', function () {
 	proclaim.isFunction(Symbol);
 });
 
-it('has correct arity', function () {
+// some old versions of chrome have arity=1, so skip this, for now
+xit('has correct arity', function () {
```
Comment (romainmenke) [inline polyfills/Symbol/polyfill.test.js:9]:
```suggestion
it.skip('has correct arity', function () {
```

**Comment by mhassan1** [comment]
@romainmenke This is ready for review.

File: polyfills/URL/_entry.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/* global require, self */
+const { URL, URLSearchParams } = require("whatwg-url");
+
+// delete these, since they should come from other polyfills
+delete URL.prototype.toJSON;
+delete URL.canParse;
+delete URL.parse;
```
Comment (romainmenke) [inline polyfills/URL/_entry.js:7]: Is it possible to strip these from the source instead of deleting them at runtime?

File: polyfills/URL/_entry.js
Comment (romainmenke) [inline polyfills/URL/_entry.js:1]: Should we have some "well known" file names for cases like these?
Or a known pattern in the naming?

If there is a pattern it is easier for people to understand the contents of a polyfill folder.

Also makes it easier to have the correct eslint rules/config for each case.

-----

At the same time I also think that this is an edge case and that we shouldn't optimize for the edge cases.

File: polyfills/URL/config.toml [hunk 3]
```diff
@@ -32,3 +49,8 @@ op_mini = "*"
 safari = "<12.0"
 ios_saf = "<12.0"
 samsung_mob = "<5.0"
+
+[install]
+module = "whatwg-url"
+postinstall = "update.task.js"
+clean = [ "polyfill.js" ]
```
Comment (romainmenke) [inline polyfills/URL/config.toml:56]:
```suggestion
clean = [ "polyfill.js", "_webidl-conversions.js" ]
```

As both files are generated, both should be removed before running `update.task.js`?

File: polyfills/URL/_webidl-conversions.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
+// this file is auto-generated by `update.task.js`
+/* global exports */
```
Comment (romainmenke) [inline polyfills/URL/_webidl-conversions.js:2]: Also fine with me to emit `/* eslint-disable */` for generated files.

**Review by romainmenke — COMMENTED** [review]
Some notes and questions, overal looks good!

The polyfill is indeed quite large, but I think that going for correctness is the right choice here.

File: polyfills/URL/_entry.js
Comment (mhassan1) [reply polyfills/URL/_entry.js:1]: I went with the `_` prefix, but I'm open to other ideas. I think each case may be different, so I'm not sure we can come up with a useful convention, without more use cases.

Comment (mhassan1) [reply polyfills/URL/_entry.js:1]: On second thought, I've removed those `_`-prefixed files from the repo. All the logic is now inside `update.task.js`.

**Comment by mhassan1** [comment]
@romainmenke I've changed the way `update.task.js` works and force-pushed. Please take a look.

File: polyfills/URL/_entry.js (hunk 2 above)
Comment (mhassan1) [reply polyfills/URL/_entry.js]: I've pushed a change to strip these out in `update.task.js`.

**Comment by mhassan1** [comment]
This PR Is blocked by https://github.com/mrhenry/polyfill-library/pull/64.

**Review by romainmenke — APPROVED** [review]
LGTM!

Thank you @mhassan1 for finding an alternative here 🙇

**Comment by romainmenke** [comment]
🤔 CI seems to be more flaky now in very old browser versions.
I suspect that the size is causing issues. (network or memory?)

Is there anything you can think of we can improve here?

**Comment by mhassan1** [comment]
> 🤔 CI seems to be more flaky now in very old browser versions. I suspect that the size is causing issues. (network or memory?)
> 
> Is there anything you can think of we can improve here?

Flaky on `URL` tests? I see timeouts are being hit [here](https://github.com/mrhenry/polyfill-library/actions/runs/10031061517/job/27722279448). Should be just increase the timeouts for the `URL` tests?

**Comment by mhassan1** [comment]
For reference, here are the largest polyfills we have:
```txt
840K	polyfills/Intl/DateTimeFormat/~timeZone/all/polyfill.js
532K	polyfills/Intl/DateTimeFormat/~timeZone/golden/polyfill.js
344K	polyfills/URL/polyfill.js
148K	polyfills/Intl/Locale/polyfill.js
144K	polyfills/String/prototype/normalize/polyfill.js
```
So `URL` is not the biggest one we have. Maybe there's something else about it that makes it slower to execute.

**Comment by romainmenke** [comment]
I think that maybe it is a really really slow parser.



Probably a combination of using many other features (e.g. `TextEncoder`, `Symbol`, iterators, `Set`, ...) and not having even the most basic optimizations (e.g. `cp2 === p(":")` vs. `cp2 === 58`).

I am concerned that url parsing is something that can happen fairly often and frequently in some client side applications. If parsing a single URL can take anywhere between ±100ms and over 2 seconds then I am sure we will be breaking some things downstream.

What options do we have here?

**Comment by mhassan1** [comment]
> I think that maybe it is a really really slow parser.

I ran the `URL` tests locally with all `detect.js` files as `false` (to force all polyfills to load in my modern browser). The test suite ran in 0.08s. Of course, my laptop is much more powerful than the BrowserStack machines; maybe this polyfill performs fine on moderately powerful machines running old browser versions.

I just ran that same test (all `detect.js` files as `false` and all `config.toml` as `*`) against a bunch of old Chrome versions on Windows 11. It seems like the really old Chrome versions (less than 46) are 10x slower than the others:
```txt
Windows 11
----------
Chrome 50: 1.40s
Chrome 49: 1.28s
Chrome 48: 1.84s
Chrome 47: 1.67s
Chrome 46: 1.73s
Chrome 45: 11.48s
Chrome 44: 11.54s
Chrome 43: 11.91s
Chrome 42: 11.96s
Chrome 41: 13.36s
Chrome 40: 16.67s

Windows XP
----------
Chrome 46: 1.23s
Chrome 45: 11.39s
```

I've narrowed down the issue to `Object.defineProperty` with `get`/`set`, which is much slower in Chrome 45 and below. Compare `for (var i = 0; i < 5000; i++) Object.defineProperty({},0,{get:function(){return 0}})` between Chrome 45 (~4s) and 46 (instant). So, I don't think the URL parser itself is slow, but it appears slow because it allocates many `ArrayBuffer` instances, which call `Object.defineProperty` many times ([link](https://github.com/mrhenry/polyfill-library/blob/e5d510bd3d5427e4ec266c4072f5ac430ae56288/polyfills/ArrayBuffer/polyfill.js#L129-L131)).

Chrome 45 doesn't get the `ArrayBuffer` polyfill, which is why it's fast when running the `URL` tests normally (without forcing it to get the polyfill, but Chrome 44 and below need the `ArrayBuffer` polyfill.

Options:
1. Increase the timeout on the `URL` tests for Chrome 44 and below (because they get the `ArrayBuffer` polyfill)
2. Drop support for Chrome 44 and below (0.03% usage total, according to https://caniuse.com/usage-table)
3. Refactor the `ArrayBuffer` polyfill to not use `Object.defineProperty` in a loop

**Comment by romainmenke** [comment]
Thank you for investigating further @mhassan1 🙇 
And sorry for the delay in picking this thread back up.

At this time I think we might need to revert the changes to `URL`.
I don't think we can drop support for the affected browser versions and I prefer dropping support for specific versions from a position we can revert from without it being a breaking change. (having to also remove shipped features like `URL.canParse` would also be breaking)

I think that the value of a working `URL` polyfill for these very old browser versions is higher than the value of `URL.parse` and `URL.canParse` in any browser version.

Do you see a path forwards that allows us to have both this polyfill for `URL` and preserve support for these old browser versions?

**Comment by romainmenke** [comment]
Upon closer inspection I am more convinced that we should revert this change.
The `URL polyfill` is currently only used for also some very old browser versions and the number of versions that could be served the updated `whatwg` polyfill and still actually need it is even smaller still.

Only:
- Chrome 45-60
- Safari 10-11
- ...

I think it is fine to have a `URL.parse` and `URL.canParse` polyfill that only work fully as expected in browsers with a native `URL` implementation that throws in all expected cases.

**Comment by romainmenke** [comment]
I've opened a PR to revert this change here: https://github.com/mrhenry/polyfill-library/pull/72

But waiting to hear your thoughts and concerns first :)
I really appreciate the time and effort you've put into this polyfill and maybe you see path forwards that I am currently overlooking.

**Comment by mhassan1** [comment]
Just to clarify: the old `URL` polyfill had incorrect behavior, regardless of `URL.parse`, so I don't think it would be better to go back to it. I think correct behavior is more important than performance.

The performance issue is caused by the frequent use of the `ArrayBuffer` polyfill, which seems to be slow. Maybe we could make it faster.

Or, we could have two `URL` polyfills that load dynamically based on the browser, but that would still mean that we are serving an incorrect polyfill to old browsers.

**Comment by mhassan1** [comment]
I've put up a PR for another proposal: remove the need for the `ArrayBuffer` polyfill for some browsers by removing the need for `TypedArray.prototype.slice` and `TypedArray.prototype.indexOf`: https://github.com/mrhenry/polyfill-library/pull/73.

**Pull request merged**